File: search/binarysearch.go
// Package search is an implementation of various search algorithm in GoLang
package search

// BinarySearch is the most efficient searching algorithm having a run-time
// complexity of O(log2 N). This algorithm works only on a sorted list of
// elements.
//
// Binary search begins by comparing the middle element of the list with the
// target element. If the target value matches the middle element, its position
// in the list is returned. If it does not match, the list is divided into two
// halves. The first half consists of the first element to middle element
// whereas the second half consists of the element next to the middle element
// to the last element.
func BinarySearch(arr []int, value int) int {
	start := 0
	end := len(arr) - 1

	for start <= end {
		median := (start + end) / 2
		if arr[median] < value {
			start = median + 1
		} else {
			end = median - 1
		}
	}

	if start == len(arr) || arr[start] != value {
		return -1
	} else {
		return start
	}
}

File: search/binary_search.py
""" Implementation of Binary Search algorithm """


def binary_search_iterative(array, item):
    """ Iterative version of Binary search
    Returns index if the item is found in the list, else returns None
    """
    start = 0
    end = len(array) - 1

    while start <= end:
        median = (start + end) // 2
        if array[median] == item:
            return median
        if array[median] < item:
            start = median + 1
        else:
            end = median - 1
    return None


def binary_search_recursive(array, start, end, item):
    """ Recursive version of binary search algorithm
    Returns index if the item is found in the list, else returns None
    """
    if end >= start:
        median = (start + end) // 2
        if array[median] == item:
            return median
        if array[median] < item:
            return binary_search_recursive(array, median+1, end, item)
        return binary_search_recursive(array, start, median-1, item)
    return None


def main():
    """ operational function """
    array = [34, 36, 46, 48, 67, 68, 69, 71, 73, 97]
    print(binary_search_iterative(array, 34))
    print(binary_search_recursive(array, 0, len(array)-1, 97))


if __name__ == "__main__":
    main()
